src/functional/program_processors/codegens/gtfn/gtfn_ir_to_gtfn_im_ir.py

```diff
@@ -0,0 +1,274 @@
+        assert offset_provider is not None
+        connectivity = _find_connectivity(node.args, offset_provider)
+        max_neighbors = connectivity.max_neighbors
+        fun, init = node.fun.args
+        args = node.args
+        # do the following transformations to the node arguments
+        # dense fields: shift(dense_f, X2Y) -> deref(shift(dense_f, X2Y, nbh_iterator)
+        # sparse_fields: sparse_f -> tuple_get(nbh_iterator, deref(sparse_f)))
+        new_args = []
+        nbh_iter = gtfn_ir_common.SymRef(id="nbh_iter")
+        for arg in args:
+            if isinstance(arg, gtfn_ir.FunCall) and arg.fun.id == "shift":
+                new_args.append(
+                    gtfn_ir.FunCall(
+                        fun=gtfn_ir_common.SymRef(id="deref"),
+                        args=[
+                            gtfn_ir.FunCall(
+                                fun=gtfn_ir_common.SymRef(id="shift"), args=arg.args + [nbh_iter]
+                            )
+                        ],
+                    )
+                )
+            if isinstance(arg, gtfn_ir_common.SymRef):
+                new_args.append(
+                    gtfn_ir.FunCall(
+                        fun=gtfn_ir_common.SymRef(id="tuple_get"),
+                        args=[
+                            nbh_iter,
+                            gtfn_ir.FunCall(fun=gtfn_ir_common.SymRef(id="deref"), args=[arg]),
+                        ],
+                    )
+                )
+
+        old_to_new_args = dict(zip([param.id for param in fun.params[1:]], new_args))
+        acc = fun.params[0]
+
+        class ReplaceArgs(NodeTranslator):
+            def visit_Expr(self, node):
+                if hasattr(node, "id") and node.id in old_to_new_args:
+                    return old_to_new_args[node.id]
+                return self.generic_visit(node)
+
+        new_fun = ReplaceArgs().visit(fun.expr)
+
+        red_idx = self.uids.sequential_id(prefix="red")
+        red_lit = gtfn_ir_common.Sym(id=f"{red_idx}")
+        self.imp_list_ir.append(InitStmt(lhs=red_lit, rhs=self.visit(init, **kwargs)))
+
+        class PlugInCurrentIdx(NodeTranslator):
+            def visit_SymRef(self, node):
+                if node.id == acc.id:
+                    return gtfn_ir_common.SymRef(id=red_idx)
+                if node.id == "nbh_iter":
+                    return self.cur_idx
+                return self.generic_visit(node)
+
+            def __init__(self, cur_idx):
+                self.cur_idx = cur_idx
+
+        for i in range(max_neighbors):
+            new_expr = PlugInCurrentIdx(cur_idx=gtfn_ir.OffsetLiteral(value=i)).visit(new_fun)
+            rhs = self.visit(new_expr, **kwargs)
+            self.imp_list_ir.append(AssignStmt(lhs=gtfn_ir_common.SymRef(id=red_idx), rhs=rhs))
+
+        return gtfn_ir_common.SymRef(id=red_idx)
+
+    def visit_FunCall(self, node: gtfn_ir.FunCall, **kwargs):
+        if isinstance(node.fun, gtfn_ir.Lambda) and any(
```

Comment:
what about symref's taking lambda's etc? Maybe excluding everything is enough
```suggestion
        if any(
```